### PR TITLE
[Mod-2145] : (iOS)Ti.Map: Support Image Overlays

### PR DESCRIPTION
**Pull request by vijaysingh-axway**
https://jira.appcelerator.org/browse/MOD-2145

### CODE REVIEW

File: apidoc/ImageOverlay.yml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        type: MapPointType
+
+      - name: bottomRight
+        summary: An object with longitude and latitude values.
+        type: MapPointType
+
+  - name: image
+    summary: The image, which needs to show on map as overlay.
+    availability: creation
+    type: [String, TiBlob]
```
Comment (hansemannn) [inline apidoc/ImageOverlay.yml]: Did you validate the new docs (by copying them into the SDK an run `node validate.js`)? `TiBlob` is not a valid type. It should be `Titanium.Blob`, so please validate it again and merge out possible issues

File: ios/Classes/TiMapImageOverlayProxy.h [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  @private
+  UIImage *image;
+}
+@end
+
+@interface TiMapImageOverlayProxy : TiProxy {
+  @private
+  UIImage *image;
+  CLLocationCoordinate2D midCoordinate;
+  MKMapRect mapRect;
```
Comment (hansemannn) [inline ios/Classes/TiMapImageOverlayProxy.h]: I started prefixing all private properties with an underscore, e.g. `_image`.

File: ios/Classes/TiMapImageOverlayProxy.m [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+- (instancetype)initWithMidCoordinate:(CLLocationCoordinate2D)midCoordinate andMapRect:(MKMapRect)mapRect
+{
+  self = [super init];
+  if (self) {
+    boundingMapRect = mapRect;
+    coordinate = midCoordinate;
+  }
+  return self;
+}
+@end
```
Comment (hansemannn) [inline ios/Classes/TiMapImageOverlayProxy.m]: Missing empty line.

File: ios/Classes/TiMapImageOverlayProxy.m [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+- (NSString *)apiName
+{
+  return @"Ti.Map.ImageOverlay";
+}
+
+- (TiMapImageOverlayRenderer *)imageOverlayRenderer
+{
+  if (imageOverlayRenderer == nil) {
+    imageOverlay = [[TiMapImageOverlay alloc] initWithMidCoordinate:midCoordinate andMapRect:mapRect];
+    imageOverlayRenderer = [[[TiMapImageOverlayRenderer alloc] initWithOverlay:imageOverlay overlayImage:image] retain];
```
Comment (hansemannn) [inline ios/Classes/TiMapImageOverlayProxy.m]: Not released. Please validate that all new objects in the map lifecycle get properly released. Also run the static analyzer to validate.

File: ios/Classes/TiMapImageOverlayProxy.m [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+  return imageOverlayRenderer;
+}
+
+#pragma mark Public APIs
+
+- (void)setImage:(id)value
+{
+  if (!value) {
+    [self throwException:@"missing required image data" subreason:nil location:CODELOCATION];
```
Comment (hansemannn) [inline ios/Classes/TiMapImageOverlayProxy.m]: `Missing required \"image\" property`. Do for other similar usages as well.

File: ios/Classes/TiMapImageOverlayProxy.m [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+  if (!value) {
+    [self throwException:@"missing required bounds coordinate data" subreason:nil location:CODELOCATION];
+  }
+
+  ENSURE_DICT(value);
+  ENSURE_ARG_COUNT(value, 2)
+  NSDictionary *topLeftDict = [value objectForKey:@"topLeft"];
+  NSDictionary *bottomRightDict = [value objectForKey:@"bottomRight"];
+
+  CLLocationDegrees topLeftLat = [TiUtils doubleValue:[topLeftDict objectForKey:@"latitude"]];
```
Comment (hansemannn) [inline ios/Classes/TiMapImageOverlayProxy.m]: Please use named parameters, e.g. `topLeftLatitude` and `topLeftDictionary`.

File: ios/Classes/TiMapImageOverlayRenderer.m [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+- (void)drawMapRect:(MKMapRect)mapRect zoomScale:(MKZoomScale)zoomScale inContext:(CGContextRef)context
+{
+  CGImageRef imageReference = self.overlayImage.CGImage;
+
+  MKMapRect theMapRect = self.overlay.boundingMapRect;
+  CGRect theRect = [self rectForMapRect:theMapRect];
+
+  CGContextScaleCTM(context, 1.0, -1.0);
+  CGContextTranslateCTM(context, 0.0, -theRect.size.height);
+  CGContextDrawImage(context, theRect, imageReference);
```
Comment (hansemannn) [inline ios/Classes/TiMapImageOverlayRenderer.m]: Do you need to release the drawing context?

File: ios/Classes/TiMapImageOverlayProxy.m (hunk 4 above)
Comment (vijaysingh-axway) [reply ios/Classes/TiMapImageOverlayProxy.m]: It's released in dealloc and validated with analyzer. One  release issue was there in TiMapImageOverlayRenderer.m. Fixed that.

File: ios/Classes/TiMapImageOverlayProxy.m (hunk 5 above)
Comment (vijaysingh-axway) [reply ios/Classes/TiMapImageOverlayProxy.m]: Done

File: ios/Classes/TiMapImageOverlayProxy.m (hunk 6 above)
Comment (vijaysingh-axway) [reply ios/Classes/TiMapImageOverlayProxy.m]: Done.

File: ios/Classes/TiMapImageOverlayRenderer.m (hunk 7 above)
Comment (vijaysingh-axway) [reply ios/Classes/TiMapImageOverlayRenderer.m]: Not required.

File: ios/Classes/TiMapImageOverlayProxy.h (hunk 2 above)
Comment (vijaysingh-axway) [reply ios/Classes/TiMapImageOverlayProxy.h]: Done

File: apidoc/ImageOverlay.yml (hunk 1 above)
Comment (vijaysingh-axway) [reply apidoc/ImageOverlay.yml]: Updated.

File: ios/manifest [hunk 8]
```diff
@@ -2,7 +2,7 @@
 # this is your module manifest and used by Titanium
 # during compilation, packaging, distribution, etc.
 #
-version: 3.0.2
+version: 3.0.3
```
Comment (hansemannn) [inline ios/manifest]: Its a feature, so it needs a minor version bump.

Comment (vijaysingh-axway) [reply ios/manifest]: You are right. Updated to 3.1.0.

**Pull request merged**